src/main.ts

```diff
@@ -182,6 +184,72 @@ export function generateFile(ctx: Context, fileDesc: FileDescriptorProto): [stri
     chunks.push(generateDataLoadersType());
   }
 
+  if (options.outputSchema) {
+    chunks.push(code`
+      export interface ProtoMetadata {
+        fileDescriptor: ${imp('IFileDescriptorProto@protobufjs/ext/descriptor')};
+        references: { [key: string]: any };
+        dependencies?: ProtoMetadata[];
+      }
+    `);
+
+    const outputFileDesc: IFileDescriptorProto = {
+      ...fileDesc,
+      sourceCodeInfo: null,
+    };
+
+    if (fileDesc.sourceCodeInfo) {
+      outputFileDesc.sourceCodeInfo = {
+        location: fileDesc.sourceCodeInfo.location.filter((loc) => loc['leadingComments'] || loc['trailingComments']),
+      } as any;
+    }
+
+    const values = [code`fileDescriptor: ${JSON.stringify(outputFileDesc)} as any`];
+
+    const references: Code[] = [];
+    visit(
+      fileDesc,
+      sourceInfo,
+      (fullName, message, sInfo) => {
+        if (options.outputEncodeMethods) {
+          references.push(code`'.${fileDesc.package}.${fullName.replace(/_/g, '.')}': ${fullName}`);
+        }
+      },
+      options,
+      (fullName, enumDesc, sInfo) => {
+        references.push(code`'.${fileDesc.package}.${fullName.replace(/_/g, '.')}': ${fullName}`);
+      }
+    );
+
+    visitServices(fileDesc, sourceInfo, (serviceDesc, sInfo) => {
+      if (options.outputClientImpl) {
+        references.push(
+          code`'.${fileDesc.package}.${serviceDesc.name.replace(/_/g, '.')}': ${serviceDesc.name}ClientImpl`
+        );
+      }
+    });
+
+    values.push(code`
+      references: {${joinCode(references, { on: ',\n' })}}
+    `);
+
+    if (fileDesc.dependency) {
+      const dependencies = fileDesc.dependency.map((dep) => {
+        const mod = dep.replace('.proto', '');
+        const localName = mod.replace(/\//g, '_') + '_protoMetadata';
+        return code`${new ImportsName(localName, './' + mod, 'protoMetadata')}`;
```

Comment:
I think this can be:

```
return code`${imp(`protoMetadata@./${mod}`)}`
```

Disclaimer that this _just_ started working in the latest ts-poet that ts-proto started using ~yesterday, where ts-poet will notice the overlapping `protoMetadata` import symbols, and auto-suffix/rename them with `import { protoMetadata as protoMetadata1 } from ./mod`;